dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/validation/PeriodTypeExtended.java

```diff
@@ -128,6 +140,18 @@ public String getDeoIds( long dataElementId, long categoryOptionComboId )
         return dataElementId + "." + categoryOptionComboId;
     }
 
+    public void setSlidingWindows( boolean slidingWindows )
+    {

```

Comment:
This part is a bit excessive. If "slidingWindowsNeeded"/"nonSlidingWindowsNeeded" are determined by "slidingWindows" only, we could have just one stateful boolean property `slidingWindows`, with getters for:

```
public boolean slidingWindowsNeeded() 
{
  return slidingWindows;
}

public boolean nonSlidingWindowsNeeded()
{
  return !slidingWindows;
}
```

I.e. better to have as few stateful properties as possible / less properties to keep in sync.

Reply:
When initializing a PeriodTypeExtended, are three possibilities: (1) all validation rule expressions in the periodType have slidingWindows = true, resulting in slidingWindowsNeeded = true and nonSlidingWindowsNeeded = false, (2) all validation rule expressions in that periodType have slidingWindows = false, resulting in slidingWindowsNeeded = false and nonSlidingWindowsNeeded = true, or (3) there are some of each, resulting in slidingWindowsNeeded = true and nonSlidingWindowsNeeded = true.

This is a performance improvement. The old code always called analytics to get the sliding windows data, and also always called analytics to get the non-sliding windows data. The new code only calls analytics for the sliding windows if needed, and only calls analytics for the non-sliding windows if needed.

Reply:
Okay makes sense.